synapse/storage/databases/main/event_federation.py

```diff
@@ -1292,6 +1292,99 @@ def _get_backfill_events(
 
         return event_id_results
 
+    @trace
+    async def record_event_failed_backfill_attempt(
+        self, room_id: str, event_id: str
+    ) -> None:
+        """
+        Record when we fail to backfill an event.
+
+        This information allows us to be more intelligent when we decide to
+        retry (we don't need to fail over and over) and we can process that
+        event in the background so we don't block on it each time.
+
+        Args:
+            room_id: The room where the event failed to backfill from
+            event_id: The event that failed to be backfilled
+        """
+        if (
+            self.database_engine.can_native_upsert
```

Comment:
we should bump our minimum supported sqlite version, so we can get rid of `can_native_upsert`.

Reply:
Discussed at https://github.com/matrix-org/synapse/pull/13589#discussion_r959840527 for a future PR

Reply:
ok; I'm just thinking it might be worth doing *before* this PR since it means you can get rid of the emulated upsert code. ymmv though.

Reply:
Going to opt for future PR. https://github.com/matrix-org/synapse/issues/12983 is a pile of work in itself deciding the on right version, documentation, making sure the right notices go out, removing other emulated upsert code, etc.

Reply:
In #13760.

Reply:
Removed the emulated upsert code. Thanks for tackling this @DMRobertson 🎉